### ISSUE TITLE
KVM Compatibility and Fix - Streaming Mode

### ISSUE THREAD

**Issue by morphinejh** [necroware/ps2-serial-mouse-adapter#9]
**This is for information and, possibly, integration.**

I have a TRENDnet TK-423K that I have tried with a few types of PS/2 to serial connections. I couldn't get any of them to work consistently. 

However, since you were kind enough to share this project publicly, I was able to adapt your code to get it all working! Turns out the KVM will not work in any mode except Streaming mode. In order to get it to work with the KVM switch that I have, I had to speed up the code execution - it seemed to get out of sync on the PS/2 side or stream too much information on the RS-232 side...

My solution was to use direct Port writes where ever possible as well as instantiate the mouse class directly in streaming mode from the start. Rather than initiating it and then resetting to put it in streaming mode. One issue with the code as it is now, is that `bool m_stream` is only set when initialized - therefore it is never fully in streaming mode after a `mouse.reset()` call. The check for:
`If (m_stream)` within `Ps2Mouse.cpp`
was always false since the constructor showed:
`Ps2Mouse::Ps2Mouse(byte clockPin, byte dataPin)
  : m_clockPin(clockPin), m_dataPin(dataPin), m_stream(false)
{}`

This essentially made it always send a command request for data, which came back all zeroes when the mouse was set to be in streaming mode. I don't know if this behavior is universal though, as I have only been able to test on the KVM switch.

The way I have written the adapted code makes it only compatible with an Arduino ProMicro due to the direct Port writes. However it could be adapted to another model easily. There are also new DEBUG statements that can be turned on/off. I did trying adapting just the new streaming mode constructor, but it didn't work correctly either. I believe due to the delays in Arduino's `digitalWrite()` and `digitalRead()`. 

Check it out the fork and let me know your thoughts:
[https://github.com/morphinejh/ps2-serial-mouse-adapter](https://github.com/morphinejh/ps2-serial-mouse-adapter)

**Comment by sweetlilmre** [comment]
@morphinejh I had a look at your fork. Direct port access is definitely the way to go here, so great addition. Why don't you raise a pull request against this repo, it would be nice to get your changes into the official source.

I have some comments on your changes, if you are interested I can fork your repo and make a pull request against it?

**Comment by morphinejh** [comment]
Not a problem. I'll check it out once you do.

**Comment by ghost** [comment]
Hi folks, 

Awesome work on the fork @morphinejh, but I was running into issues with a KVM as well, a D-Link DKVM-8E, it sometimes has a bit of a brain fart and send the cursor flying across the screen and clicking randomly. This is likely a timing issue or something, but I just didn't have time to look into it, so I took the error checking route. 
When this happens, most of the time, there are parity errors on the packets or the overflow bits on the X and/or Y movement are set, so I finished implementing the parity check stub in the code and added an X/Y overflow check.
This made the adapter usable for me, but I would still occasionally have the cursor teleport across the screen, so I added a simple mechanism to trim out the movements that are above a configurable limit, this feature is enabled by a jumper(D13) and the limit value is set with a pot (A1) that can be adjusted  on-the-fly. I found in my tests that a movement >128 is very likely an error, so this solved all the issues for me. 
I forked your fork and made a pull request with my code, hope it helps.

**Comment by perchrh** [comment]
Nice! The additional trimpot and jumper, are they an addition to the PCB? If so, could you share updated kicad files as well? :)

**Comment by ghost** [comment]
> Nice! The additional trimpot and jumper, are they an addition to the PCB? If so, could you share updated kicad files as well? :)

No, sorry, kicad is still a bit intimidating to me, that is why I made the feature entirely optional, so that it didn't invalidate the current board design, that bit of code for the movement trimming can be ignored and is just there in case someone **really** needs it. 
I promised myself I would I would learn kicad, it's on the list....

**Comment by ghost** [comment]
Hi @morphinejh,

I was still running into a lot of issues with my KVM, I implemented a bunch of error handling to address most of the issues, so that I would just get a small hiccup in the mouse movement when my KVM has a brain fart.
I tuned as much as I could all the timings, but I just couldn't get everything fixed. The challenge for me was keeping both the serial and PS/2 baud rates locked in, so in the end I resorted to implementing the serial communication over the HW Serial on the TX pin on my arduino nano.
**That worked perfectly, since the Serial message queue was now handled in HW, I could focus only on the PS/2 baud rate.**
 It works absolutely flawlessly now but you do kind of have to give up on the Serial for debugging messages (Check how I handled DEBUG...). 
To test my version you should just have to bridge pin 7 (D4) to the TX pin, if it all goes well you will see the TX LED blinking with your mouse movements!
Somehow I totally missed @sweetlilmre 's PR, I am definitely going to merge  in my code! Love the wheel mouse support! But for now I am just going to enjoy some Simcity 2000. 

Cheers!
